# 1- Análise Exploratória e Pré-processamento

In [67]:
import pandas as pd

In [68]:
df = pd.read_csv("data/akc-data-latest.csv")
df.head(3)

Unnamed: 0                                        description  \
0     Affenpinscher  The Affen’s apish look has been described many...   
1      Afghan Hound  The Afghan Hound is an ancient breed, his whol...   
2  Airedale Terrier  The Airedale Terrier is the largest of all ter...   

                                 temperament popularity  min_height  \
0        Confident, Famously Funny, Fearless        148       22.86   
1  Dignified, Profoundly Loyal, Aristocratic        113       63.50   
2               Friendly, Clever, Courageous         60       58.42   

   max_height  min_weight  max_weight  min_expectancy  max_expectancy  ...  \
0       29.21    3.175147    4.535924            12.0            15.0  ...   
1       68.58   22.679619   27.215542            12.0            15.0  ...   
2       58.42   22.679619   31.751466            11.0            14.0  ...   

  grooming_frequency_value  grooming_frequency_category shedding_value  \
0                      0.6    2-3 Times a Week Brushing            0.6   
1                      0.8               Daily Brushing            0.2   
2                      0.6    2-3 Times a Week Brushing            0.4   

   shedding_category energy_level_value  energy_level_category  \
0           Seasonal                0.6       Regular Exercise   
1         Infrequent                0.8              Energetic   
2         Occasional                0.6       Regular Exercise   

  trainability_value  trainability_category demeanor_value  demeanor_category  
0                0.8          Easy Training            1.0           Outgoing  
1                0.2        May be Stubborn            0.2         Aloof/Wary  
2                1.0        Eager to Please            0.8           Friendly  

[3 rows x 21 columns]

Como 21 colunas é um número relativamente alto, seria interessante encontrar maneiras de reduzir a dimensionalidade dos dados

In [69]:
len(df)

277

277 é um número baixo de dados, pode ser que seja necessário fazer algum tipo de data augmentation

In [70]:
# Exemplo de linha completa, com todas as colunas e seus respectivos valores
for column in df.columns:
    print(column + " - " + str(df.iloc[0][column]))

Unnamed: 0 - Affenpinscher
description - The Affen’s apish look has been described many ways. They’ve been called “monkey dogs” and “ape terriers.” The French say diablotin moustachu (“mustached little devil”), and Star Wars fans argue whether they look more like Wookies or Ewoks.Standing less than a foot tall, these sturdy terrier-like dogs approach life with great confidence. “This isn’t a breed you train,” a professional dog handler tells us, “He’s like a human. You befriend him.” The dense, harsh coat is described as “neat but shaggy” and comes in several colors; the gait is light and confident. They can be willful and domineering, but mostly Affens are loyal, affectionate, and always entertaining. Affen people say they love being owned by their little monkey dogs. The Affenpinscher: loyal, curious, and famously amusing; this almost-human toy dog is fearless out of all proportion to his size. As with all great comedians, it’s the Affenpinscher’s apparent seriousness of purpose that

In [71]:
df.describe()

min_height  max_height  min_weight  max_weight  min_expectancy  \
count  277.000000  277.000000  275.000000  275.000000      274.000000   
mean    44.225801   52.720588   17.888858   27.291416       11.306569   
std     14.238298   15.885454   12.290600   19.061416        1.817949   
min     12.700000   17.780000    0.000000    0.000000        0.000000   
25%     33.020000   38.100000    8.164663   12.927383       10.000000   
50%     45.085000   53.340000   15.875733   24.947580       12.000000   
75%     55.880000   66.040000   22.679619   34.019428       12.000000   
max     76.200000   88.900000   68.038855  108.862169       16.000000   

       max_expectancy  grooming_frequency_value  shedding_value  \
count      274.000000                270.000000      257.000000   
mean        13.832117                  0.425926        0.529183   
std          2.016668                  0.198306        0.189068   
min          0.000000                  0.200000        0.200000   
25%         13.000000                  0.200000        0.400000   
50%         14.000000                  0.400000        0.600000   
75%         15.000000                  0.600000        0.600000   
max         19.000000                  1.000000        1.000000   

       energy_level_value  trainability_value  demeanor_value  
count          271.000000          253.000000      252.000000  
mean             0.712915            0.624506        0.620635  
std              0.168927            0.247271        0.201713  
min              0.200000            0.200000        0.200000  
25%              0.600000            0.400000        0.400000  
50%              0.600000            0.600000        0.600000  
75%              0.800000            0.800000        0.800000  
max              1.000000            1.000000        1.000000

In [72]:
print(df["group"].value_counts())

group
Foundation Stock Service    68
Hound Group                 32
Sporting Group              32
Terrier Group               31
Working Group               31
Herding Group               30
Toy Group                   21
Non-Sporting Group          21
Miscellaneous Class         11
Name: count, dtype: int64


O nosso objetivo é tentar predizer o "group" de um objeto (supervisionado) e tentar agrupar objetos em torno dos grupos (não supervisionado)

### 1.0 - As colunas Unnamed: 0, description e popularity são únicas para cada elemento do dataset, logo não ajudarão no treinamento e podem ser removidas

In [73]:
df["Unnamed: 0"].value_counts()

Unnamed: 0
Affenpinscher                   1
Papillon                        1
Petit Basset Griffon Vendéen    1
Peruvian Inca Orchid            1
Perro de Presa Canario          1
                               ..
English Cocker Spaniel          1
English Foxhound                1
English Setter                  1
English Springer Spaniel        1
Yorkshire Terrier               1
Name: count, Length: 277, dtype: int64

In [74]:
df["description"].value_counts()

description
Poodles come in three size varieties: Standards should be more than 15 inches tall at the shoulder; Miniatures are 15 inches or under; Toys stand no more than 10 inches. All three varieties have the same build and proportions. At dog shows, Poodles are usually seen in the elaborate Continental Clip. Most pet owners prefer the simpler Sporting Clip, in which the coat is shorn to follow the outline of the squarely built, smoothly muscled body.Forget those old stereotypes of Poodles as sissy dogs. Poodles are eager, athletic, and wickedly smart “real dogs” of remarkable versatility. The Standard, with his greater size and strength, is the best all-around athlete of the family, but all Poodles can be trained with great success. Whether Standard, Miniature, or Toy, and either black, white, or apricot, the Poodle stands proudly among dogdom’s true aristocrats. Beneath the curly, low-allergen coat is an elegant athlete and companion for all reasons and seasons.                    

OBS: A descrição as vezes é bem semelhante para animais de mesma raça, sendo completamente igual apenas uma vez para o Poodle (Miniature) e Poodle (Standard) (no Poodle (Toy) ela já varia um pouco), porém muitas vezes isso não acontece e as descrições acabam sendo strings bem longas, portanto achei que seria complicado tentar manter algum tipo de relação entre as descrições para ajudar na predição e achei melhor apenas desconsiderar a coluna mesmo

In [75]:
df["popularity"].value_counts()

popularity
7      3
of     3
17     2
140    2
133    2
      ..
176    1
67     1
52     1
188    1
10     1
Name: count, Length: 191, dtype: int64

In [76]:
for _, element in df.loc[df["popularity"] == "7"].iterrows():
    print(element["group"])

Non-Sporting Group
Non-Sporting Group
Toy Group


In [77]:
for _, element in df.loc[df["popularity"] == "of"].iterrows():
    print(element["group"])

Hound Group
Sporting Group
Working Group


In [78]:
for _, element in df.loc[df["popularity"] == "17"].iterrows():
    print(element["group"])

Herding Group
Working Group


OBS: O popularity as vezes se repete, porém essas repetições não aparentam ter uma relação muito direta com o "group" que é o nosso label, achei melhor tirar junto com os outros 2

In [79]:
df_clean = df.drop(columns=["Unnamed: 0", "description", "popularity"])

### 1.1 - Para todos os campos de baixo, é redundante manter os 2 por isso vou manter apenas o value e não o category (uma vez que já está normalizado de 0 - 1)

In [80]:
print(df_clean["grooming_frequency_value"].value_counts())
print()
print(df_clean["grooming_frequency_category"].value_counts())

grooming_frequency_value
0.4    119
0.2     75
0.6     50
0.8     18
1.0      8
Name: count, dtype: int64

grooming_frequency_category
Weekly Brushing              119
Occasional Bath/Brush         75
2-3 Times a Week Brushing     50
Daily Brushing                18
Specialty/Professional         8
Name: count, dtype: int64


In [81]:
print(df_clean["shedding_value"].value_counts())
print()
print(df_clean["shedding_category"].value_counts())

shedding_value
0.6    125
0.4     59
0.2     37
0.8     30
1.0      6
Name: count, dtype: int64

shedding_category
Seasonal      125
Occasional     59
Infrequent     37
Regularly      30
Frequent        6
Name: count, dtype: int64


In [82]:
print(df_clean["energy_level_value"].value_counts())
print()
print(df_clean["energy_level_category"].value_counts())

energy_level_value
0.6    118
0.8     92
1.0     41
0.4     19
0.2      1
Name: count, dtype: int64

energy_level_category
Regular Exercise          118
Energetic                  92
Needs Lots of Activity     41
Calm                       19
Couch Potato                1
Name: count, dtype: int64


In [83]:
print(df_clean["trainability_value"].value_counts())
print()
print(df_clean["trainability_category"].value_counts())

trainability_value
0.6    77
0.4    66
1.0    50
0.8    39
0.2    21
Name: count, dtype: int64

trainability_category
Agreeable          77
Independent        66
Eager to Please    50
Easy Training      39
May be Stubborn    21
Name: count, dtype: int64


In [84]:
print(df_clean["demeanor_value"].value_counts())
print()
print(df_clean["demeanor_category"].value_counts())

demeanor_value
0.8    77
0.6    75
0.4    73
1.0    19
0.2     8
Name: count, dtype: int64

demeanor_category
Friendly                   77
Alert/Responsive           75
Reserved with Strangers    73
Outgoing                   19
Aloof/Wary                  8
Name: count, dtype: int64


In [85]:
df_clean = df_clean.drop(columns=["grooming_frequency_category", "shedding_category", "energy_level_category", "trainability_category", "demeanor_category"])

### 1.2 - Analisar se vai ser necessário tirar Temperament

In [91]:
df["temperament"].value_counts()

temperament
Friendly, Smart, Willing to Please    3
Friendly, Alert, Intelligent          2
Loyal, Intelligent, Active            2
Loyal, Independent, Calm              2
Loyal, Alert, Intelligent             2
                                     ..
Intelligent, Lively, Athletic         1
Energetic, Merry, Responsive          1
Affectionate, Gentle, Sociable        1
Friendly, Mellow, Merry               1
Affectionate, Sprightly, Tomboyish    1
Name: count, Length: 267, dtype: int64

### 1.3 - Correlação

In [89]:
for column in df_clean.columns:
    print(column + " - " + str(df_clean.iloc[0][column]))

temperament - Confident, Famously Funny, Fearless
min_height - 22.86
max_height - 29.21
min_weight - 3.17514659
max_weight - 4.535923700000001
min_expectancy - 12.0
max_expectancy - 15.0
group - Toy Group
grooming_frequency_value - 0.6
shedding_value - 0.6
energy_level_value - 0.6
trainability_value - 0.8
demeanor_value - 1.0
